lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1314 @@
+		let mut constructor =
+			Self { state_machine, channel_id, inputs_to_contribute, outputs_to_contribute };
+		let message_send = if is_initiator {
+			match constructor.maybe_send_message() {
+				Ok(msg_send) => Some(msg_send),
+				Err(_) => {
+					debug_assert!(
+						false,
+						"We should always be able to start our state machine successfully"
+					);
+					None
+				},
+			}
+		} else {
+			None
+		};
+		(constructor, message_send)
+	}
+
+	fn maybe_send_message(&mut self) -> Result<InteractiveTxMessageSend, AbortReason> {
+		// We first attempt to send inputs we want to add, then outputs. Once we are done sending
+		// them both, then we always send tx_complete.
+		if let Some((serial_id, input, prevtx)) = self.inputs_to_contribute.pop() {
+			let msg = msgs::TxAddInput {
+				channel_id: self.channel_id,
+				serial_id,
+				prevtx,
+				prevtx_out: input.previous_output.vout,
+				sequence: input.sequence.to_consensus_u32(),
+			};
+			do_state_transition!(self, sent_tx_add_input, &msg)?;
+			Ok(InteractiveTxMessageSend::TxAddInput(msg))
+		} else if let Some((serial_id, output)) = self.outputs_to_contribute.pop() {
+			let msg = msgs::TxAddOutput {
+				channel_id: self.channel_id,
+				serial_id,
+				sats: output.value,
+				script: output.script_pubkey,
+			};
+			do_state_transition!(self, sent_tx_add_output, &msg)?;
+			Ok(InteractiveTxMessageSend::TxAddOutput(msg))
+		} else {
+			let msg = msgs::TxComplete { channel_id: self.channel_id };
+			do_state_transition!(self, sent_tx_complete, &msg)?;
+			Ok(InteractiveTxMessageSend::TxComplete(msg))
+		}
+	}
+
+	pub fn handle_tx_add_input(
+		&mut self, msg: &msgs::TxAddInput,
+	) -> Result<InteractiveTxMessageSend, AbortReason> {
+		do_state_transition!(self, received_tx_add_input, msg)?;
+		self.maybe_send_message()
+	}
+
+	pub fn handle_tx_remove_input(
+		&mut self, msg: &msgs::TxRemoveInput,
+	) -> Result<InteractiveTxMessageSend, AbortReason> {
+		do_state_transition!(self, received_tx_remove_input, msg)?;
+		self.maybe_send_message()
+	}
+
+	pub fn handle_tx_add_output(
+		&mut self, msg: &msgs::TxAddOutput,
+	) -> Result<InteractiveTxMessageSend, AbortReason> {
+		do_state_transition!(self, received_tx_add_output, msg)?;
+		self.maybe_send_message()
+	}
+
+	pub fn handle_tx_remove_output(
+		&mut self, msg: &msgs::TxRemoveOutput,
+	) -> Result<InteractiveTxMessageSend, AbortReason> {
+		do_state_transition!(self, received_tx_remove_output, msg)?;
+		self.maybe_send_message()
+	}
+
+	pub fn handle_tx_complete(
+		&mut self, msg: &msgs::TxComplete,
+	) -> Result<(Option<InteractiveTxMessageSend>, Option<Transaction>), AbortReason> {
+		do_state_transition!(self, received_tx_complete, msg)?;
+		match &self.state_machine {
+			StateMachine::ReceivedTxComplete(_) => {
+				let msg_send = self.maybe_send_message()?;
+				let negotiated_tx = match &self.state_machine {
+					StateMachine::NegotiationComplete(s) => Some(s.0.clone()),
```

Comment:
Consider using `Option<&Transaction>` to avoid cloning if possible at the call site.